pkg/broker/work_engine.go

```diff
@@ -29,13 +29,13 @@ type Work interface {
 
 // WorkEngine - a new engine for doing work.
 type WorkEngine struct {
-	topics map[WorkTopic]chan JobMsg
-	bufsz  int
+	subscribers map[WorkTopic][]WorkSubscriber
+	jobs        map[string]chan JobMsg
 }
 
 // NewWorkEngine - creates a new work engine
-func NewWorkEngine(bufferSize int) *WorkEngine {
-	return &WorkEngine{topics: make(map[WorkTopic]chan JobMsg), bufsz: bufferSize}
+func NewWorkEngine() *WorkEngine {
+	return &WorkEngine{jobs: make(map[string]chan JobMsg), subscribers: map[WorkTopic][]WorkSubscriber{}}
```

Comment:
again, why is the channel a non buffered channel?

Reply:
As currently we are handing off the messages immediately and asynchronously we should always be able to read from and send to the channel so in this case there is no benefit to a buffered channel.  

Where we may benefit from a buffered channel is if we used it in conjunction with a ```sync.WaitGroup``` 

Something like:

```
  wg := sync.WaitGroup{}

  for msg := range engine.jobs[token] {
      wg.Add(len(engine.subscribers))
      for _, sub := range engine.subscribers[topic] {
          go sub.Notify(msg, wg) // Subscriber will call Done()
      }
      wg.Wait() // wait for all subscribers to be done
  }
```

This design would allow each of the subscribers to receive the messages at the same time, but force us to wait until they are all complete before accepting the next message. In this case having a buffer would be valuable as it would allow the Job to continue to send messages even if our subscribers were running a little slow. 
The question really is if there is value in waiting for all the subscribers to finish before accepting the next message. 
I am leaning towards yes on this due to the fact that state will be changed and modified and there is no guarantee on that the order a set of go routines were created will also be the order they are executed in.